compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -368,10 +368,15 @@ pub fn check_unsafety<'tcx>(tcx: TyCtxt<'tcx>, def: ty::WithOptConstParam<LocalD
         return;
     }
 
-    // Closures are handled by their parent function
+    // Closures are handled by their parent function...
     if tcx.is_closure(def.did.to_def_id()) {
-        tcx.ensure().thir_check_unsafety(tcx.hir().local_def_id_to_hir_id(def.did).owner);
-        return;
+        // ...unless they occur inside an ADT definition, e.g. in the
+        // length part of an array (issue #85871)
+        let owner = tcx.hir().local_def_id_to_hir_id(def.did).owner;
+        if !tcx.type_of(owner).is_adt() {
```

Comment:
I'm not sure if this is adequate. For example, what about `type Foo = [...; || ...];`?

We probably want to identify the cases where we *should* check the owner instead and enumerate those.

Reply:
You are right, `.is_adt()` is not comprehensive enough (although it is not a regression from the current behavior either).

I have used `.maybe_body_owned_by(...).is_some()` now, which should be more robust, and expanded the test case; could you have another look?